### PR TITLE
Remove dictionary from ILoaderOptions

### PR DESCRIPTION
**Pull request by ChumpChief**
Effectively this change would clarify "ILoaderOptions is for known options on the Loader layer only".

Currently, ILoaderOptions permits setting any option, as it's a dictionary type.  This random dictionary functionality is used by:
1. client: IClient (a long-standing secret handshake here)
2. enableOfflineLoad: boolean (a newer secret handshake)
3. A few DDS-level features, mostly sequence stuff and attribution stuff.

This change would make the first two explicit but marked deprecated (assuming their secrecy means we don't want anyone to use them).  It would disallow the third category at the typing layer but avoid making functional changes immediately.

The third category seems an inappropriate use, since it's referring to a container-global options object for what are generally per-DDS settings.  This seems to have been convenient for testing purposes, but seems dangerous for a customer to use (e.g. imagine two DataObjects fighting over their desired configs on the global object).  Perhaps this was even done accidentally, as it's not obvious that dataStoreRuntime.options is secretly referring to a container-global options object - it really does sound like it should be options for just that dataStoreRuntime.  My hope is that we can stem further usage with this change and work towards replacing the existing usage with something more appropriate.

Eventually this may allow us to remove the plumbing of the ILoaderOptions through the layers, since the specified properties aren't super interesting to the other layers.

Acknowledge that this is a breaking change (to typing, not to functionality at runtime), and I need to remember what we're doing with those now.

### CODE REVIEW

File: packages/dds/cell/src/test/cell.spec.ts [hunk 1]
```diff
@@ -24,7 +24,10 @@ function createConnectedCell(
 ): ISharedCell {
 	// Create and connect a second SharedCell.
 	const dataStoreRuntime = new MockFluidDataStoreRuntime();
-	dataStoreRuntime.options = options ?? dataStoreRuntime.options;
+	// TODO this option shouldn't live here - this options object is global to the container
```
Comment (anthony-murphy) [inline packages/dds/cell/src/test/cell.spec.ts:27]: i think the first thing is to change the type of  dataStoreRuntime be something other than ILoaderOptions, probably Record<string | number, any> to start. that would also get rid of all the casting this change introduces

Comment (anthony-murphy) [reply packages/dds/cell/src/test/cell.spec.ts:27]: i'm also not positive that the province of this object actually comes all the way from the loader. the type might just be a lie

Comment (ChumpChief) [reply packages/dds/cell/src/test/cell.spec.ts:27]: I left this annoying intentionally - my thought here was to discourage folks from writing new values to options (with this PR as-is, they'll get an error).  This is with the goal of deterring new (possibly dangerous) usage.

I think the likely end state being that the dataStoreRuntime.options member goes away entirely.

Comment (ChumpChief) [reply packages/dds/cell/src/test/cell.spec.ts:27]: > i'm also not positive that the province of this object actually comes all the way from the loader. the type might just be a lie

The loader options get shallow-copied per-Container - so shallow modifications are global to the Container, but deep modifications of true loader options (e.g. individual IClient properties) are global to the Loader.  But yeah they get plumbed all the way through.

**Comment by ChumpChief** [comment]
@Abe27342 particularly interested to get your thought on attribution and other DDS usage, and what alternatives we might consider.

**Comment by Abe27342** [comment]
> @Abe27342 particularly interested to get your thought on attribution and other DDS usage, and what alternatives we might consider.

I looked into this a while back. IMO all the DDS-level "options" should live somewhere near `IFluidDataStoreRegistry` in the API surface. #15397 (never merged) took a stab at that. One problem in the realm it doesn't address is that we will need clean ways to declare migrations for options which need consensus across clients. At one point Tony suggested more heavy tweaks to the data store registry pattern which might allow that kind of flow, but whatever we do will need some design work.

I know one of the stated goals of this PR is trying to discourage adding further options to ILoaderOptions which shouldn't be there, but I don't love that this also obfuscates how apps using Fluid can set these options. If we do end up with this intermediate state, I think we'll need some clear partner communication around what we're planning.

**Comment by ChumpChief** [comment]
> I know one of the stated goals of this PR is trying to discourage adding further options to ILoaderOptions which shouldn't be there, but I don't love that this also obfuscates how apps using Fluid can set these options. If we do end up with this intermediate state, I think we'll need some clear partner communication around what we're planning.

My goal is not to obfuscate, but ultimately to make it illegal and stop passing loader options through to the runtime at all.  If the registry-based approach works out (or another alternative) then the communication to partners would be to use that instead.  Are you aware of customers using loader options to control these features already today?

**Comment by Abe27342** [comment]
Yeah, if plan is to check something like this in alongside an alternative way for partners to set these options then awesome, I'd love that.

I don't know exactly how our partners set all of these options, but several techniques that I've seen which *don't* involve passing thru loader actually have correctness issues. All the other methods I've seen basically rely on mutation of some options objects, and it's a crapshoot whether that applies before or after the option is read for whatever it affects.

E.g. one of the merge-tree flags we've since removed was mutated on construction of the dataStoreRuntime in a `hasInitialized` call, but when I looked into it I discovered that the option was inconsistent depending on whether you were creating or loading the DDS, and also the mutating code never ran for the summarizer client at all.

I'd guess that there are some partners which do pass in some of these as loader options.

Obviously not a great state to be in.

**Comment by ChumpChief** [comment]
All this sounds good to me I think, though I'd also emphasize the importance of not just the mechanism (options vs. config) but more particularly the site of setting (i.e. at what layer).  Anything relating to the specific contents of the runtime/DDS layer seems inappropriate to be configured from the loader layer (or the consumer of the loader layer).  Conceptually it even seems odd that the consumer of a Loader would try to load some container code and say "by the way, if you happen to have any merge-trees in there I have an opinion about your snapshot chunk size".

I'd argue that these should probably be set in the runtime layer, probably by the author of the runtime factory code or DataObject code.  And even if the Loader does have some unique information that should figure in ("I'm using a driver that likes 5kb chunks") then this info should instead be injected for the runtime to respond to ("As the runtime author, I'll use that preferred chunk size to tweak my merge-tree behaviors").

> I'd guess that there are some partners which do pass in some of these as loader options.

Since these are all hidden/undocumented, I don't think a partner could have reasonably found these and taken a dependency unless we personally advised them to?  And as you mention many are not safe to vary across loads, making it unlikely that customers with varying deployments of their loader-consuming code could have safely used them.

**Comment by Abe27342** [comment]
> I'd argue that these should probably be set in the runtime layer, probably by the author of the runtime factory code or DataObject code. 

Yeah, definitely agreed. Just want to emphasize that we don't have a safe way to do that today.

File: packages/dds/cell/src/test/cell.spec.ts (hunk 1 above)
Comment (ChumpChief) [reply packages/dds/cell/src/test/cell.spec.ts:27]: I'm looking at an option to split the typing (maybe this is what you originally meant with your comment) - ILoaderOptions with only known types for the Loader constructor param, but everywhere else gets JUST the record-like.  Which then would point to a followup change that just severs the connection at the loader/runtime boundary, with the runtime creating an empty object as its starting options and getting nothing from the loader.

Comment (ChumpChief) [reply packages/dds/cell/src/test/cell.spec.ts]: Updated in latest!

File: packages/common/container-definitions/src/runtime.ts [hunk 2]
```diff
@@ -123,7 +123,9 @@ export interface IBatchMessage {
  * @alpha
  */
 export interface IContainerContext {
-	readonly options: ILoaderOptions;
+	// Used to be ILoaderOptions, this is staging for eventual removal.
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	readonly options: Record<string | number, any>;
```
Comment (anthony-murphy) [inline packages/common/container-definitions/src/runtime.ts:128]: could this be a separate PR that decouples context from loader options. that would be non-breaking

Comment (anthony-murphy) [reply packages/common/container-definitions/src/runtime.ts:128]: that could include container runtime, and data store runtime too, bascially decouple the lower layer first

File: packages/common/container-definitions/src/loader.ts [hunk 3]
```diff
@@ -533,10 +533,8 @@ export interface IHostLoader extends ILoader {
 /**
  * @public
  */
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
 export type ILoaderOptions = {
```
Comment (anthony-murphy) [inline packages/common/container-definitions/src/loader.ts:537]: i like removing it, but i think its used today, so we need a transition path, i don't think we can just unilaterally remove

File: packages/loader/container-loader/src/container.ts [hunk 4]
```diff
@@ -2032,13 +2034,12 @@ export class Container
...
 		containerId: string,
-		options?: ILoaderOptions,
+		loaderOptionsClient?: IClient,
 		clientDetailsOverride?: IClientDetails,
 	): IClient {
-		const loaderOptionsClient = structuredClone(options?.client);
 		const client: IClient =
 			loaderOptionsClient !== undefined
-				? (loaderOptionsClient as IClient)
+				? structuredClone(loaderOptionsClient)
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/container.ts:2042]: i know in a previous bump loop has problems with this, is it currently used elsewhere? is it safe to use?

**Review by anthony-murphy — APPROVED** [review]

File: packages/common/container-definitions/src/loader.ts [hunk 5]
```diff
@@ -533,10 +533,8 @@ export interface IHostLoader extends ILoader {
 /**
  * @public
```
Comment (anthony-murphy) [inline packages/common/container-definitions/src/loader.ts:534]:
```suggestion
 * @alpha
```

Comment (anthony-murphy) [reply packages/common/container-definitions/src/loader.ts:534]: this only had to be public as it propagated through the layers

File: packages/common/container-definitions/src/loader.ts [hunk 6]
```diff
@@ -531,12 +531,10 @@ export interface IHostLoader extends ILoader {
 }
 
 /**
- * @public
+ * @alpha
```
Comment (Josmithr) [inline packages/common/container-definitions/src/loader.ts:534]: Docs while we're here? 😁

File: packages/common/container-definitions/src/runtime.ts (hunk 2 above)
Comment (Josmithr) [inline packages/common/container-definitions/src/runtime.ts:128]: Especially now that we've lost type information here, let's document what these options are for.
